pip/index.py

```diff
@@ -330,13 +335,16 @@ def mkurl_pypi_url(url):
                 ) or req.url_name
 
         if url_name is not None:
-            locations = [
-                mkurl_pypi_url(url)
-                for url in self.index_urls] + self.find_links
-        else:
-            locations = list(self.find_links)
+            return [mkurl_pypi_url(url) for url in self.index_urls]
+        return []
+
+    def _find_all_versions(self, req):
```

Comment:
most importantly,  that this method does NOT filter versions found by the version specifer, only by project name at this point.   

an idea for another simple refactor, would be to only pass around the `project_name` through these methods, if that's all that is being used.